File: 1.Data structure/3.Stack/주식 가격/주식가격.java
import java.util.Stack;

class Solution {
    public int[] solution(int[] prices) {
        int len = prices.length;
        int[] answer = new int[len];
        Stack<Integer> stack = new Stack<>();
        int i,temp;
        stack.push(0);   
        for (i = 1; i < len; i++){
            while(!stack.empty() && prices[stack.peek()] > prices[i]){
                temp = stack.pop();
                answer[temp] = i-temp;
            }
            stack.push(i);
        }
        while (!stack.empty()){
            temp = stack.pop();
            answer[temp] = len - temp-1;
        }
        
        return answer;
    }
}

// 일반 풀이
        // int len = prices.length;
        // int[] answer = new int[len];
        // int i;
        // for (i = 0; i < len-1; i++){
        //     int k = i;
        //     while (prices[i] <= prices[k]){
        //         k += 1;
        //         if (k >= len - 1){
        //             break;
        //         }
        //     }
        //     answer[i] = k-i;
        // }

File: 1.Data structure/3.Stack/주식가격.py
# stack 사용 
def solution(prices):
    answer = [0] * len(prices)
    stack = []
    for i in range(0,len(prices)):
        while stack != [] and stack[-1][0] > prices[i]:
            temp = stack.pop()[1]
            answer[temp] = i-temp
        else:
            stack.append([prices[i],i])
    for i in stack:
        answer[i[1]] = len(prices) - i[1]-1
    return answer

# 원래 코드, 테스트 3 222.59, 19.3mb
    # answer =[]
    # for i in range(0,len(prices)-1):
    #     k = i
    #     while prices[i] <= prices[k]:
    #         k +=1
    #         if k >= len(prices)-1:
    #             break
    #     answer.append(k-i)
    # answer.append(0)